Pd upgrade

liv20 opened the pull request:
In acnportal.acnsim.base, changed the from_json and to_json serialization functions to match with the pandas v2 API. Checked that "python -m unittest tests.test_json_io" still passed.

sunash commented on acnportal/acnsim/base.py:
```diff
@@ -14,7 +14,7 @@
 import pkg_resources
 import pandas
-from pandas.io.common import stringify_path, get_handle, get_filepath_or_buffer
+from pandas.io.common import stringify_path, get_handle, _get_filepath_or_buffer
```
I'm a bit concerned about backward compatibility here. Users running pandas 1 will get an error here. Would it be possible to keep this working for pandas 1 by checking pandas install version but throw a deprecation warning if the user is using pandas 1?

sunash commented on acnportal/acnsim/tests/test_interface.py:
```diff
@@ -272,9 +273,8 @@ def setUp(self) -> None:
-        self.network.magnitudes = np.ones((4, 1))
-        self.network.constraint_index = ["C1", "C2", "C3", "C4"]
+        for i, station_id in enumerate(self.network.station_ids):
+            self.network.add_constraint(Current(station_id), 1, f"C{i+1}")
```
On my end, add_constraint seems to break in pandas 2 due to DataFrames not having an append method.

sunash reviewed (changes requested):
Thanks for the PR! Updating to pandas 2 has been on the to-do list for a while. A few modifications needed:
- DataFrame.append doesn't seem to be a method in pandas 2, so need to update ChargingNetwork.add_constraint() to do the same thing with pandas 2 methods.
- Backwards compatibility with pandas 1 is desirable for at least a few months after we merge in these changes
- Some serialization tests fail due to a version conflict between the currently running version of pandas (2) and version stored at the time of serialization. I'm open to thoughts on how to resolve this, but would like all tests to pass before merging.

If this isn't an urgent update, I can look at these issues more in-depth and contribute in about 3 weeks. Otherwise I can keep monitoring this PR but won't have time to code anything up myself.

liv20 commented:
Thanks for the review! Just submitted another pull request. It'd be best if we could have pandas 2 ready by 10/11 (~3 weeks).

- I tried to impose backwards compatibility by checking the pandas version number
- updated the append method for pandas v2 (I believe the relevant test programs have passed)
- Also ran black, only did the 2 files I changed but can re-run it on the entire package if we'd like
- I'm unsure about serialization since it may not work to serialize in pandas 1 and deserialize in pandas 2. Perhaps we could just re-run the serialization and store them as json using pandas 2? On my end, running "python -m unittest tests.test_json_io" passes.

liv20 commented:
Also, another concern is that the old "get_filepath_or_buffer" function is preceded by an underscore "_get_filepath_or_buffer", which means pandas may get rid of it in future versions. Open to thoughts on the best way to resolve it (could write our version of the pandas's code since the functions "_get_filepath_or_buffer" calls are not underscored)

sunash reviewed (commented):
Almost LGTM, just two minor comments. Good work!

sunash commented:
By the way, the serialization/verison conflict thing turned out to be an issue on my end. As soon as I reinstall acnportal, the more relaxed pandas requirements kick in and all the tests pass for me as well. So no worries there.

liv20 commented on acnportal/acnsim/network/charging_network.py:
```diff
@@ -251,7 +251,27 @@ def add_constraint(
+        if pd.__version__ > "1.4.0":
+            if len(constraint_frame) == 0:
+                constraint_frame_ = current.to_frame().T
+                for col in constraint_frame_:
```
Agreed. The point was to add columns in the empty DataFrame onto the appended frame. The first line should be ```constraint_frame``` and the second should be ```_constraint_frame```.

liv20 commented on acnportal/acnsim/base.py:
```diff
@@ -14,7 +14,17 @@
+    )
+    from pandas.io.common import stringify_path, get_handle, get_filepath_or_buffer
+else:
+    from pandas.io.common import stringify_path, get_handle, _get_filepath_or_buffer
```
Just did - I took pandas code for ```_get_filepath_or_buffer``` and adapted it to our needs. From ```from_json```, it seems like we don't plan to test URLs. On my end, the serialization/deserialization tests past. Let me know if you have any feedback.

sunash commented:
Hmm, I'm seeing some merge conflicts, could you resolve them and I'll take a look after that?

liv20 commented:
OK, let me know if you have any feedback!

sunash commented:
LGTM, just merged. Thanks!
